share/scripts/sdcdb.tcl

```diff
@@ -0,0 +1,795 @@
+        if {$tmp ne {} && $tmp ne $filename} {
+            error "'$funcname' not found in file '$filename' according to database"
+        }
+    }
+    if {$record eq {}} {
+        error "'$funcname' not found in file database"
+    }
+    return $record
+}
+
+proc search_file {file line} {
+    set tmp [file rootname $file]
+    if {[file extension $file] eq ".c"} {
+        set arrayname c_${tmp}2addr
+        variable c_${tmp}2addr
+    } else {
+        set arrayname a_${tmp}2addr
+        variable a_${tmp}2addr
+    }
+    if {![info exists $arrayname]} {
+        error "'$file': source file not found ($arrayname)"
+    }
+    set result [array get $arrayname $line]
+    if {$result eq {}} {
+        error "line $line not found in file '$file'"
+    }
+    lindex $result 1
+}
+
+# recurring pattern
+proc map_source_params {pos fun1 fun2 fun3 args} {
+    # parameter pattern 1: file:beginLine
+    set pattern1 {([^:]+):(\d+)}
+    # parameter pattern 2: file:functionName
+    set pattern2 {([^:]+):(\S+)}
+    # parameter pattern 3: functionName
+    set pattern3 {(\S+)}
+    set match [regexp -inline $pattern1 $pos]
+    if {[llength $match] == 3} {
+        return [$fun1 {*}[lrange $match 1 end] {*}$args]
+    }
+    set match [regexp -inline $pattern2 $pos]
+    if {[llength $match] == 3} {
+        return [$fun2 {*}[lrange $match 1 end] {*}$args]
+    }
+    set match [regexp -inline $pattern3 $pos]
+    if {[llength $match] == 2} {
+        return [$fun3 {*}[lrange $match 1 end] {*}$args]
+    }
+    # ignore the rest
+}
+
+proc sdcdb_map {pos} {
+    map_source_params $pos map_fline map_ffunc map_func
+}
+
+proc map_fline {filename linenum} {
+    search_file $filename $linenum
+}
+
+proc map_ffunc {filename funcname} {
+    search_func $filename $funcname
+}
+
+proc map_func {funcname} {
+    search_func {} $funcname
+}
+
+proc sdcdb_break {pos {cond {}} {cmd {sdcdb info -break}}} {
+    map_source_params $pos break_fline break_ffunc break_func $cond $cmd
+}
+
+proc break_fline {filename linenum cond cmd} {
+    set record [search_file $filename $linenum]
+    if {$record eq {}} {
+        error "address not found"
+    } else {
+        debug breakpoint create -address [lindex $record 1] -condition $cond -command $cmd
+    }
+}
+
+proc break_ffunc {filename funcname cond cmd} {
+    lassign [search_func $filename $funcname] start {}
+    if {$start eq {}} {
+        error "function not found"
+    } else {
+        debug breakpoint create -address $start -condition $cond -command $cmd
+    }
+}
+
+proc break_func {funcname cond cmd} {
+    lassign [search_func {} $funcname] start {}
+    if {$start eq {}} {
+        error "function not found"
+    }
+    debug breakpoint create -address $start -condition $cond -command $cmd
+}
+
+proc check {} {
+    variable old_PC
+    return [reg PC] != $old_PC
+}
+
+proc print_status {file line} {
+    puts "file: $file:$line, position: 0x[h [reg PC]]"
```

Comment:
This is less clear. This command is meant to be used interactively (less useful to be used by a script). But who knows what crazy ideas our users have for scripts. Maybe it is useful after all to also capture this output in a script. And there's also no real downside to returning vs printing. Both end up in the console by default.

Reply:
It can return instead of print as long as `update_set` prints it, since `update_step` runs in background and cannot return anything.

Reply:
Good point, I missed that.

And such functions are not meaningful to be called from other scripts anyway, So directly using `puts` here seems appropriate.